libs/tiled-bitmap/src/cmyklayeroperations.cc

```diff
@@ -0,0 +1,331 @@
+/*
+ * Scroom - Generic viewer for 2D data
+ * Copyright (C) 2009-2020 Kees-Jan Dijkzeul
+ *
+ * SPDX-License-Identifier: LGPL-2.1
+ */
+
+#include <scroom/layeroperations.hh>
+
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <glib.h>
+
+#include <sstream>
+
+#include <boost/shared_ptr.hpp>
+#include <boost/utility.hpp>
+
+#include <scroom/bitmap-helpers.hh>
+#include <scroom/cairo-helpers.hh>
+#include <scroom/unused.hh>
+
+#include <boost/math/special_functions/round.hpp>
+#include <iostream>
+
+////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+  boost::shared_ptr<unsigned char> shared_malloc(size_t size)
+  {
+    return boost::shared_ptr<unsigned char>(static_cast<unsigned char*>(malloc(size)), free);
+  }
+}
+
+////////////////////////////////////////////////////////////////////////
+// OperationsCMYK
+
+LayerOperations::Ptr OperationsCMYK::create(int bps)
+{
+  return Ptr(new OperationsCMYK(static_cast<uint16_t>(bps)));
+}
+
+OperationsCMYK::OperationsCMYK(uint16_t bps_)
+  : bps(bps_)
+{
+}
+
+int OperationsCMYK::getBpp()
+{
+  // CMYK has 4 channels -> 4 samples per pixel.
+  return this->bps * 4;
+}
+
+// From https://www.pagetable.com/?p=23#comment-1140
+inline uint16_t DivideBy255(uint16_t value)
+{
+  return static_cast<uint16_t>((value + 1 + (value >> 8)) >> 8);
```

Comment:
Premature optimization?

Reply:
Moving the division by 255 to a specialised function causes a performance increase of a little over 3% (3.35s -> 3.24s) when loading a 14391 * 6232 8bps TIFF image. Recent improvements have decreased this number though. In a previous version, the performance increase due to this function was almost 15%. As such, I wouldn't call this optimisation premature.

However, whether the 3% load time decrease is worth the reduced readibility is a less objective matter. In my opinion, the readability of the code doesn't suffer enough to change this. If you disagree and think this should not be in a separate function, I'm happy to change it.

Reply:
For only 3%, I would not choose to use constructs like this one.